trie/ctrie/ctrie.go

```diff
@@ -0,0 +1,571 @@
+	}
+	return found.(*sNode).value, true
+}
+
+// inserted creates a new L-node with the added entry.
+func (l *lNode) inserted(entry *entry) *lNode {
+	return &lNode{l.Add(&sNode{entry})}
+}
+
+// removed creates a new L-node with the entry removed.
+func (l *lNode) removed(e *entry) *lNode {
+	idx := l.FindIndex(func(sn interface{}) bool {
+		return bytes.Equal(e.key, sn.(*sNode).key)
+	})
+	if idx < 0 {
+		return l
+	}
+	nl, _ := l.Remove(uint(idx))
+	return &lNode{nl}
+}
+
+// length returns the L-node list length.
+func (l *lNode) length() uint {
+	return l.Length()
+}
+
+// branch is either an iNode or sNode.
+type branch interface{}
+
+// entry contains a Ctrie entry, which is also a technique used to cache the
+// hashcode of the key.
+type entry struct {
+	key   []byte
+	hash  uint32
+	value interface{}
+}
+
+// sNode is a singleton node which contains a single key and value.
+type sNode struct {
+	*entry
+}
+
+// New creates an empty Ctrie, defaulting to FNV-1a for key hashing. Use
+// SetHash to change the hash function.
+func New() *Ctrie {
+	root := &iNode{main: &mainNode{cNode: &cNode{}}}
+	return &Ctrie{root: root, h: fnv.New32a()}
+}
+
+// SetHash sets the hash function used by the Ctrie. Existing entries are not
+// rehashed when this is set, so this should be called on a newly created
+// Ctrie.
+func (c *Ctrie) SetHash(hash hash.Hash32) {
+	c.hMu.Lock()
+	c.h = hash
+	c.hMu.Unlock()
+}
+
+// Insert adds the key-value pair to the Ctrie, replacing the existing value if
+// the key already exists.
+func (c *Ctrie) Insert(key []byte, value interface{}) {
+	c.insert(&entry{
+		key:   key,
+		hash:  c.hash(key),
+		value: value,
+	})
+}
+
+// Lookup returns the value for the associated key or returns false if the key
+// doesn't exist.
+func (c *Ctrie) Lookup(key []byte) (interface{}, bool) {
+	return c.lookup(&entry{key: key, hash: c.hash(key)})
+}
+
+// Remove deletes the value for the associated key, returning true if it was
+// removed or false if the entry doesn't exist.
+func (c *Ctrie) Remove(key []byte) (interface{}, bool) {
+	return c.remove(&entry{key: key, hash: c.hash(key)})
+}
+
+func (c *Ctrie) insert(entry *entry) {
+	rootPtr := (*unsafe.Pointer)(unsafe.Pointer(&c.root))
+	root := (*iNode)(atomic.LoadPointer(rootPtr))
+	if !iinsert(root, entry, 0, nil) {
+		c.insert(entry)
+	}
+}
+
+func (c *Ctrie) lookup(entry *entry) (interface{}, bool) {
+	rootPtr := (*unsafe.Pointer)(unsafe.Pointer(&c.root))
+	root := (*iNode)(atomic.LoadPointer(rootPtr))
+	result, exists, ok := ilookup(root, entry, 0, nil)
+	for !ok {
+		return c.lookup(entry)
+	}
+	return result, exists
+}
+
+func (c *Ctrie) remove(entry *entry) (interface{}, bool) {
+	rootPtr := (*unsafe.Pointer)(unsafe.Pointer(&c.root))
+	root := (*iNode)(atomic.LoadPointer(rootPtr))
+	result, exists, ok := iremove(root, entry, 0, nil)
+	for !ok {
+		return c.remove(entry)
+	}
+	return result, exists
+}
+
+func (c *Ctrie) hash(k []byte) uint32 {
+	c.hMu.Lock()
```

Comment:
So looking at the rest of the code, I'm guessing this will be the primary bottleneck as almost everything calls hash and will be linearized right here.  A suggestion might be to ask the consumer to pass in a hash factory or use a default if none passed in and then fill a ringbuffer with some number of hashers.  Puts to the buffer will never block and gets will be as fast as any hasher can hash.  Found that sort of model worked well in the PALM btree.